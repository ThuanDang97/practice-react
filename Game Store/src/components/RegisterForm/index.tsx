import { Link } from 'react-router-dom'
import { useForm, SubmitHandler, Controller } from 'react-hook-form'
import { Box, Typography, useTheme } from '@mui/material'

// Constants
import {
  INVALID_EMAIL,
  INVALID_PASSWORD,
  REGEX_EMAIL,
  REQUIRED,
} from '@constants/index'

// Components
import Button from '@components/Button'
import { InputStyled } from '@components/Input/InputStyled'

// Types
import { IAccountRegister } from '@self-types/index'

export interface IRegisterFormProps {
  handleSubmitForm: (data: IAccountRegister) => void
}

const RegisterForm = ({ handleSubmitForm }: IRegisterFormProps) => {
  const theme = useTheme()
  const {
    formState: { errors },
    handleSubmit,
    control,
  } = useForm<IAccountRegister>()

  const onSubmit: SubmitHandler<IAccountRegister> = (data) => {
    handleSubmitForm(data)
  }

  return (
    <Box
      minWidth={350}
      maxWidth={450}
      padding={4}
      component="form"
      onSubmit={handleSubmit(onSubmit)}
      display="flex"
      justifyContent="center"
      flexDirection="column"
      alignItems="initial"
      width="100%"
      bgcolor={theme.palette.common.white}
      borderRadius="5px"
    >
      <Typography
        color={theme.palette.common.black}
        fontWeight={700}
        fontSize={28}
        variant="h3"
        marginBottom={3}
        textAlign="center"
      >
        Register
      </Typography>
      <Box display="flex" gap={1} alignItems="initial">
        <Box display="flex" flexDirection="column" width="100%">
          <Controller
            name="firstName"
            control={control}
            defaultValue=""
            rules={{ required: { value: true, message: REQUIRED } }}
            render={({ field }) => (
              <InputStyled
                error={!!errors.firstName}
                helperText={errors.firstName && errors.firstName.message}
                color="primary"
                focused
                variant="standard"
                type="text"
                placeholder="First name"
                {...field}
              />
            )}
          />
        </Box>
        <Box display="flex" flexDirection="column" width="100%">
          <Controller
            name="lastName"
            control={control}
            defaultValue=""
            rules={{ required: { value: true, message: REQUIRED } }}
            render={({ field }) => (
              <InputStyled
                error={!!errors.lastName}
                helperText={errors.lastName && errors.lastName.message}
                color="primary"
                focused
                variant="standard"
                type="text"
                placeholder="Last name"
                displayline="none"
                {...field}
              />
            )}
          />
        </Box>
      </Box>
      <Controller
        name="email"
        control={control}
        defaultValue=""
        rules={{
          required: { value: true, message: REQUIRED },
          pattern: { value: REGEX_EMAIL, message: INVALID_EMAIL },
        }}
        render={({ field }) => (
          <InputStyled
            error={!!errors.email}
            helperText={errors.email && errors.email.message}
            color="primary"
            focused
            variant="standard"
            type="text"
            placeholder="Email"
            {...field}
          />
        )}
      />
      <Controller
        name="password"
        control={control}
        defaultValue=""
        rules={{
          required: { value: true, message: REQUIRED },
          minLength: { value: 8, message: INVALID_PASSWORD },
        }}
        render={({ field }) => (
          <InputStyled
            error={!!errors.password}
            helperText={errors.password && errors.password.message}
            color="primary"
            focused
            variant="standard"
            type="password"
            placeholder="Password"
            {...field}
          />
        )}
      />
      <Button mtop={50} color="primary" type="submit">
        Register
      </Button>
      <Box
        display="flex"
        justifyContent="center"
        alignItems="center"
        marginTop={1}
      >
        <Typography color={theme.palette.common.black} fontSize={14}>
          If you already have an account?
        </Typography>
        <Link to="/login">
          <Typography
            paddingLeft={0.5}
            color={theme.palette.info.main}
            fontSize={14}
            fontWeight={500}
          >
            Login now
          </Typography>
        </Link>
      </Box>
    </Box>
  )
}

export default RegisterForm
