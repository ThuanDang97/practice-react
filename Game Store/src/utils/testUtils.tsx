import { ReactNode } from 'react'
import { render } from '@testing-library/react'
import { ThemeProvider } from '@mui/material/styles'
import { BrowserRouter as Router } from 'react-router-dom'

// Themes
import theme from '@self-types/Themes'

// Contexts
import { AuthProvider } from '@contexts/AuthProvider'
import { CartProvider } from '@contexts/CartProvider'

export const customRender = (component: ReactNode) => {
  return render(
    <ThemeProvider theme={theme}>
      <AuthProvider>
        <CartProvider>
          <Router>{component}</Router>
        </CartProvider>
      </AuthProvider>
    </ThemeProvider>,
  )
}

export * from '@testing-library/react'
export { customRender as render }
