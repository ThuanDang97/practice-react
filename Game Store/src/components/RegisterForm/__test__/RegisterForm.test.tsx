import { fireEvent, screen } from '@testing-library/react'

// Utils
import { render } from '@utils/testUtils'

// Components
import RegisterForm from '@components/RegisterForm'

describe('Register Form component', () => {
  const handleSubmit = jest.fn()
  const props = {
    firstName: 'Le',
    lastName: 'Nhan',
    email: 'nhanle@gmail.com',
    password: '123456789',
  }
  it('should render Register form', () => {
    const { container } = render(
      <RegisterForm handleSubmitForm={handleSubmit} />,
    )

    expect(container).toMatchSnapshot()
  })

  it('should call onChange input first name', () => {
    render(<RegisterForm handleSubmitForm={handleSubmit} />)
    const input = screen.getByPlaceholderText('Last name')

    fireEvent.change(input, {
      target: { value: props.firstName },
    })

    expect(input).toHaveValue(props.firstName)
  })

  it('should call onChange input last name', () => {
    render(<RegisterForm handleSubmitForm={handleSubmit} />)
    const input = screen.getByPlaceholderText('First name')

    fireEvent.change(input, {
      target: { value: props.lastName },
    })

    expect(input).toHaveValue(props.lastName)
  })

  it('should call onChange input email', () => {
    render(<RegisterForm handleSubmitForm={handleSubmit} />)
    const input = screen.getByPlaceholderText('Email')

    fireEvent.change(input, {
      target: { value: props.email },
    })

    expect(input).toHaveValue(props.email)
  })

  it('should call onChange input password', () => {
    render(<RegisterForm handleSubmitForm={handleSubmit} />)
    const input = screen.getByPlaceholderText('Password')

    fireEvent.change(input, {
      target: { value: props.password },
    })

    expect(input).toHaveValue(props.password)
  })
})
