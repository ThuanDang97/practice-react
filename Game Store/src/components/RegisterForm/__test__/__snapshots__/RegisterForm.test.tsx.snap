// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Form component should render Register form 1`] = `
<div>
  <form
    class="MuiBox-root css-opvnev"
  >
    <h3
      class="MuiTypography-root MuiTypography-h3 css-1nyssq8-MuiTypography-root"
    >
      Register
    </h3>
    <div
      class="MuiBox-root css-1w7d0io"
    >
      <div
        class="MuiBox-root css-1ofqig9"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1pcdsgt-MuiFormControl-root-MuiTextField-root"
        >
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary Mui-focused MuiInputBase-formControl css-1ptx2yq-MuiInputBase-root-MuiInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
              id=":r0:"
              name="firstName"
              placeholder="First name"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root css-1ofqig9"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-inevag-MuiFormControl-root-MuiTextField-root"
          displayline="none"
        >
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary Mui-focused MuiInputBase-formControl css-1ptx2yq-MuiInputBase-root-MuiInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
              id=":r1:"
              name="lastName"
              placeholder="Last name"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root css-1pcdsgt-MuiFormControl-root-MuiTextField-root"
    >
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary Mui-focused MuiInputBase-formControl css-1ptx2yq-MuiInputBase-root-MuiInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
          id=":r2:"
          name="email"
          placeholder="Email"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root css-1pcdsgt-MuiFormControl-root-MuiTextField-root"
    >
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary Mui-focused MuiInputBase-formControl css-1ptx2yq-MuiInputBase-root-MuiInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
          id=":r3:"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-78dfj0-MuiButtonBase-root-MuiButton-root"
      hovercolor="white"
      mtop="50"
      tabindex="0"
      type="submit"
      width="100%"
    >
      Register
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <div
      class="MuiBox-root css-1xxl6si"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-b5y2m8-MuiTypography-root"
      >
        If you already have an account?
      </p>
      <a
        href="/login"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-ai1txx-MuiTypography-root"
        >
          Login now
        </p>
      </a>
    </div>
  </form>
</div>
`;
